File: data-structs/hash-bst-rbt/hashtable.py
'''Author: Christian Garcia
Contents: Hash Table'''

import random

class HashNode:
    def __init__(self, key, data):
        self.key = key
        self.data = data
        self.link = None


class HashTable:

    hash_array: HashNode
    capacity: int
    size: int

    def __init__(self):
        self.capacity = 10
        self.size = 0
        self.hash_array = [None] * self.capacity

    def isEmpty(self):
        return (self.size() == 0)

    def hashCode(self, key):
        return key % self.capacity

    def getIndex(self, key: int):
        code = self.hashCode(key)
        index = code % self.capacity

    def insert(self, key, data):
        self.size += 1
        i = self.hashCode(key)        
        if self.hash_array[i] == None:
            self.hash_array[i] = HashNode(key, data)
        else:
            curr = self.hash_array[i]
            while curr.link != None:
                if curr.key == key:
                    curr.key = key
                    curr.data = data
                    return
                curr = curr.link
                curr.link = HashNode(key, data)

    def remove(self, key):
        self.size -= 1
        i = self.hashCode(key)
        curr = prev = self.hash_array[i]
        if not curr:
            return
        if curr.key == key:
            self.hash_array[i] = curr.link
        else:
            curr = curr.link
            while curr:
                if curr.key == key:
                    prev.link = curr.link
                    break
                else:
                    curr, prev = curr.link, prev.link

    def printTable(self):
        for i in range(self.capacity):
            if self.hash_array[i] == None:
                print("key:   | array[ ]:  ")
            else:    
                print("key: " + str(self.hash_array[i].key) + " | array[" + str(i) +"]: " + str(self.hash_array[i].data))


'''Main'''
hash_size = 0
hash_table = HashTable()
for i in range(hash_table.capacity):
    hash_table.insert(i, random.randint(1, 101))
print("The size of the hash table is: " + str(hash_table.size))
print("\nHere is the original table: ")
hash_table.printTable()
hash_table.remove(4)
print("\nThe size of the hash table is: " + str(hash_table.size))
print("\nHere is the modified table: ")
hash_table.printTable()                

File: data-structs/hash-bst-rbt/hashtable.cpp
#include <iostream>
#include <list>
using namespace std;

class HashTable {
  int capacity;
  list<int> *table;

public:
  HashTable(int c);
  void insert(int k, int d);
  void deleteNode(int k);
  void displayTable();

  int hashCode(int k) { return k % capacity; }

  int checkNum(int n) {
    if (n == 0 || n == 1) {
      return 0;
    }
    for (int i = 2; i < n / 2; i++) {
      if (n % i == 0) {
        return 0;
      }
    }
    return 1;
  }

  int getNum(int n) {
    if (n % 2 == 0) {
      n++;
    }
    while (!checkNum(n)) {
      n += 2;
    }
    return n;
  }
};

HashTable::HashTable(int c) {
  int size = getNum(c);
  this->capacity = size;
  table = new list<int>[capacity];
}

void HashTable::insert(int k, int d) {
  int i = hashCode(k);
  table[i].push_back(d);
}

void HashTable::deleteNode(int k) {
  int index = hashCode(k);
  list<int>::iterator i;
  for (i = table[index].begin(); i != table[index].end(); i++) {
    if (*i == k) {
      break;
    }
  }
  if (i != table[index].end()) {
    table[index].erase(i);
  }
}

void HashTable::displayTable() {
  for (int i = 0; i < capacity; i++) {
    cout << "table[" << i << "]";
    for (auto x : table[i])
      cout << " --> " << x;
    cout << endl;
  }
}

int main() {
  int key[] = {231, 321, 212, 321, 433, 262};
  int data[] = {123, 432, 523, 43, 423, 111};
  int size = sizeof(key) / sizeof(key[0]);

  HashTable h(size);

  for (int i = 0; i < size; i++)
    h.insert(key[i], data[i]);

  h.deleteNode(12);
  h.displayTable();
}